### PR TITLE
[C-2338][C-2487][C-2489] Add Related Artists on Profile (web)

### PR DESCRIPTION
**Pull request by amendelsohn**
### Description

- Adds related artists preview box on profile
	- on click, opens related artists modal
	- only shows if we found related artists that are not the fallback top artists we use for suggested follows
- Migrates `artist-recommendations` slice -> new `related-artists` slice
	- uses createEntityAdapter for cleaner slice
	- keyed by `artistId` so states are independent across pages
	- add state to track whether we are using the fallback `topArtists`

image

### Dragons

- `user-list/related-artists` slice is separate from `related-artists` slice
- there may be bugs with scrolling and infinite loading of related artists in the user list modal

### How Has This Been Tested?

Local web build

### Feature Flags ###

`RELATED_ARTISTS_ON_PROFILE_ENABLED`

### CODE REVIEW

File: packages/common/src/store/user-list/related-artists/reducers.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const userListReducer = UserListReducerFactory.createReducer({
+  tag: RELATED_ARTISTS_USER_LIST_TAG,
+  pageSize: 15
+})
+
+const initialState = {
+  id: null
+}
+
+const relatedArtistsPageReducer = createReducer<
```
Comment (sliptype) [inline packages/common/src/store/user-list/related-artists/reducers.ts:20]: Is there a reason we aren't using `createSlice` here? For consistency with the other user-lists?

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/amendel-related-artists-web

File: packages/web/src/assets/img/iconUserGroup.svg [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (sliptype) [inline packages/web/src/assets/img/iconUserGroup.svg:1]: Do we need this svg in both web and stems?

File: packages/web/src/pages/profile-page/ProfilePageProvider.tsx [hunk 3]
```diff
@@ -976,7 +976,7 @@ function makeMapStateToProps() {
       playing: getPlaying(state),
       buffering: getBuffering(state),
       pathname: getLocationPathname(state),
-      relatedArtists: getRelatedArtists(state, {
+      relatedArtists: selectRelatedArtistsUsers(state, {
```
Comment (sliptype) [inline packages/web/src/pages/profile-page/ProfilePageProvider.tsx:979]: I like the `select` prefix! 🙌

**Review by sliptype — APPROVED** [review]
Great work!

File: packages/common/src/store/user-list/related-artists/reducers.ts (hunk 1 above)
Comment (amendelsohn) [reply packages/common/src/store/user-list/related-artists/reducers.ts:20]: Yeah every other user-list used createReducer. I'm sure it'd work either way though. Thoughts?

Comment (sliptype) [reply packages/common/src/store/user-list/related-artists/reducers.ts:20]: I would prefer createSlice bc the other ones will be migrated eventually (hopefully)

File: packages/common/src/store/ui/related-artists/sagas.ts [hunk 4]
```diff
@@ -38,16 +38,19 @@ export function* fetchRelatedArtists(action: Action) {
...
+        showingTopArtists = true
       }
     }
     if (filteredArtists.length > 0) {
-      const relatedArtistIds: ID[] = yield call(cacheUsers, filteredArtists)
-      yield put(
-        artistRecommendationsActions.fetchRelatedArtistsSucceeded({
-          userId,
-          relatedArtistIds
+      const relatedArtistIds: ID[] = yield* cacheUsers(filteredArtists)
```
Comment (dylanjeffers) [inline packages/common/src/store/ui/related-artists/sagas.ts]: is it possible to not explicitly type here?

File: packages/common/src/store/ui/related-artists/sagas.ts [hunk 5]
```diff
@@ -57,7 +60,7 @@ export function* fetchRelatedArtists(action: Action) {
 function* fetchTopArtists() {
   yield* waitForRead()
   const apiClient = yield* getContext('apiClient')
-  const currentUserId: ID = yield select(getUserId)
+  const currentUserId: ID | null = yield* select(getUserId)
```
Comment (dylanjeffers) [inline packages/common/src/store/ui/related-artists/sagas.ts]: here as well

File: packages/common/src/store/ui/related-artists/sagas.ts [hunk 6]
```diff
@@ -57,7 +60,7 @@ export function* fetchRelatedArtists(action: Action) {
...
   const topArtists: User[] = yield apiClient.getTopArtists({
```
Comment (dylanjeffers) [inline packages/common/src/store/ui/related-artists/sagas.ts]: and yes would be most ideal to push the types as far as possible

File: packages/common/src/store/ui/related-artists/selectors.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export const selectRelatedArtists = (state: CommonState, props: { id: ID }) => {
+  return selectRelatedArtistsById(state, props.id)
+}
+
+export const selectRelatedArtistsUsers = createDeepEqualSelector(
+  [selectRelatedArtistIds, getUsers],
+  (relatedArtistIds, users) => {
+    if (!relatedArtistIds) return []
```
Comment (dylanjeffers) [inline packages/common/src/store/ui/related-artists/selectors.ts:29]: i think we want to put [] outside so it's stable

File: packages/common/src/store/ui/related-artists/selectors.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    const relatedArtistsPopulated = relatedArtistIds
+      .map((id) => {
+        if (id in users) return users[id]
```
Comment (dylanjeffers) [inline packages/common/src/store/ui/related-artists/selectors.ts:32]: i think users[id] is fine, cause we will removeNullable after? `removeNullable` actually removes anything that is falsy

File: packages/mobile/src/screens/profile-screen/ArtistRecommendations/ArtistRecommendations.tsx [hunk 9]
```diff
@@ -102,7 +102,7 @@ export const ArtistRecommendations = (props: ArtistRecommendationsProps) => {
   })
 
   const suggestedArtists = useProxySelector(
```
Comment (dylanjeffers) [inline packages/mobile/src/screens/profile-screen/ArtistRecommendations/ArtistRecommendations.tsx:104]: need proxy selector? just cause it's expensive and really needed when rerenders happen all the time

File: packages/web/src/common/store/user-list/related-artists/sagas.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+function* errorDispatcher(error: Error) {
+  const id = yield* select(getId)
+  if (id) {
+    yield* put(getRelatedArtistsError(id, error.message))
+  }
+}
+
+function* getRelatedArtists(currentPage: number, pageSize: number) {
+  const id: number | null = yield* select(getId)
```
Comment (dylanjeffers) [inline packages/web/src/common/store/user-list/related-artists/sagas.ts:60]: explicit typing

**Review by dylanjeffers — APPROVED** [review]
stellar job!

File: packages/web/src/components/related-artists/RelatedArtists.module.css [hunk 11]
```diff
@@ -0,0 +1,12 @@
+.container {
+  margin-top: 32px;
```
Comment (dylanjeffers) [inline packages/web/src/components/related-artists/RelatedArtists.module.css:2]: lets use `var(--unit-8)`

File: packages/common/src/store/user-list/related-artists/reducers.ts (hunk 1 above)
Comment (amendelsohn) [reply packages/common/src/store/user-list/related-artists/reducers.ts:20]: resolved to stick with the implementaiton consistent with the parallel user-list slices, but filing a work item to refactor the whole thing

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/amendel-related-artists-web

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/amendel-related-artists-web

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/amendel-related-artists-web

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/amendel-related-artists-web

**Pull request merged**